tools/tpm2_hash.c

```diff
@@ -152,7 +152,7 @@ static bool on_option(char key, char *value) {
 bool tpm2_tool_onstart(tpm2_options **opts) {
 
     static struct option topts[] = {
-        {"hierarchy", required_argument, NULL, 'H'},
+        {"hierarchy", optional_argument, NULL, 'H'},
```

Comment:
Thanks for the pointer, I'm not very familiar with getopt but by the time I'd finished opening the pull request I was beginning to lean to **-H-** for consistency with the later changes, I'll make this change.

Reply:
I didn't know it has an optional_argument parameter, so I had to go check it out.

Reply:
Actually, this is just plain wrong for the change as implemented. Right the -H option *requires* an argument. If H isn't passed we default to the owner hierarchy, if it is passed we expect a hierarchy value to be defined.

I think the right fix here is to change this back to required_argument. We *could* take -H- and implicitly set owner hierarchy, but that's already the default so I'm not sure it makes much sense.

Reply:
On top of that we are confusing `-H` when it's a hierarchy versus handle.

When it's specifying a handle, it should be required and have `-` argument to mean pick a persistent handle for me.

When it's the hierarchy, it should be for specifying the hierarchy and default to owner.

Reply:
That's how this series implements things, for the most part. The only place we vary slightly is that for *tpm2_evictcontrol* if the handle passed via **-H** is a transient handle and no **-p** is specified we pick a persistent slot to persist the transient handle at.

We could:
- leave this as implemented
- require **-H-** be passed to pick a persistent slot
- optionally take **-H-** to pick a persistent slot, otherwise fall back on the logic as implemented (picking a persistent slot to use when **tpm2_evictcontrol -H** is invoked with a transient handle)

Reply:
Well `H` or `c` (which c is just a context blob where it get's the value of `H` from) need to be specified no matter what. So I think we just leave this as implemented, or make `-p-` mean pick a slot, thoughts?

Reply:
I'm leaning to leave this as implemented.